File: build_core.py
"""Упаковка расчётного ядра в zip для загрузки в Pyodide.

Запускать после любого изменения в ``core/``:
    python build_core.py

Результат кладётся в ``web/public/core/core.zip`` — оттуда его забирает
браузер и распаковывает в виртуальную ФС Pyodide.
"""

from __future__ import annotations

import pathlib
import zipfile

ROOT = pathlib.Path(__file__).parent
CORE_DIR = ROOT / "core"
OUTPUT = ROOT / "web" / "public" / "core" / "core.zip"


def build() -> None:
    """Собрать core/ в zip, пропуская кеш и байт-код."""
    OUTPUT.parent.mkdir(parents=True, exist_ok=True)
    files = [
        f
        for f in CORE_DIR.rglob("*.py")
        if "__pycache__" not in f.parts
    ]
    with zipfile.ZipFile(OUTPUT, "w", zipfile.ZIP_DEFLATED) as archive:
        for file in files:
            archive.write(file, file.relative_to(ROOT).as_posix())

    size_kb = OUTPUT.stat().st_size / 1024
    print(f"Упаковано {len(files)} файлов -> {OUTPUT} ({size_kb:.1f} КБ)")


if __name__ == "__main__":
    build()

File: web/build-core.mjs
/**
 * Упаковка расчётного ядра (Python core/) в zip для Pyodide.
 *
 * Node-версия build_core.py — нужна для деплоя: хостинг (Cloudflare
 * Pages и т.п.) запускает при сборке Node, а не Python, поэтому zip
 * собирается здесь и встроен в `npm run build`.
 *
 * Результат: web/public/core/core.zip — браузер распаковывает его в
 * виртуальную ФС Pyodide.
 */

import { readdirSync, statSync, mkdirSync, writeFileSync } from "node:fs";
import { join, relative, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import AdmZip from "adm-zip";

const scriptDir = dirname(fileURLToPath(import.meta.url));
// Скрипт лежит в web/, корень проекта — на уровень выше.
const root = join(scriptDir, "..");
const coreDir = join(root, "core");
const outPath = join(root, "web", "public", "core", "core.zip");

/** Рекурсивно собрать все .py, пропуская __pycache__. */
function collectPyFiles(dir) {
  const result = [];
  for (const entry of readdirSync(dir)) {
    if (entry === "__pycache__") continue;
    const full = join(dir, entry);
    if (statSync(full).isDirectory()) {
      result.push(...collectPyFiles(full));
    } else if (entry.endsWith(".py")) {
      result.push(full);
    }
  }
  return result;
}

function build() {
  const files = collectPyFiles(coreDir);
  const zip = new AdmZip();
  for (const file of files) {
    // Путь внутри архива — относительно корня, с прямыми слэшами.
    const archivePath = relative(root, file).split("\\").join("/");
    zip.addLocalFile(file, dirname(archivePath));
  }
  mkdirSync(dirname(outPath), { recursive: true });
  zip.writeZip(outPath);

  const sizeKb = (statSync(outPath).size / 1024).toFixed(1);
  console.log(`Упаковано ${files.length} файлов -> core.zip (${sizeKb} КБ)`);
}

build();
